src/zope/interface/registry.py

```diff
@@ -70,6 +72,53 @@ def __delitem__(self, component):
 
 class _UtilityRegistrations(object):
 
+    # Strong reference {id(components): _UtilityRegistrations}
+    _regs_for_components = {}
+    # Weak reference {id(components): components}, used for cleanup.
+    _weakrefs_for_components = {}
+
+    @classmethod
+    def for_components(cls, components):
+        # We manage these utility/subscription registrations as associated
+        # objects with a weakref to avoid making any changes to
+        # the pickle format. They are keyed off the id of the component because
+        # Components subclasses are not guaranteed to be hashable.
+        key = id(components)
```

Comment:
The initial version of this patch did use a WeakKeyDictionary, but [it turned out](https://github.com/zopefoundation/zope.interface/issues/53) that there are Components subclasses that aren't hashable (the aforementioned Pyramid registry that also subclasses dict). So we had to [revert that and use id instead](https://github.com/zopefoundation/zope.interface/pull/54).